experimental/go/internalgsdk.go

```diff
@@ -0,0 +1,262 @@
+	return &internalGsdk{
+		currentGameState:              "",
+		transitionToActiveEvent:       make(chan interface{}),
+		transitionToActiveEventClosed: false,
+		signalHeartbeatEvent:          make(chan interface{}),
+		configuration:                 nil,
+		configMap:                     nil,
+		healthCallback:                nil,
+		maintenanceCallback:           nil,
+		shutdownCallback:              nil,
+		connectedPlayers:              nil,
+		initialPlayers:                nil,
+		startInternalRan:              false,
+		startInternalMutex:            &sync.Mutex{},
+	}
+}
+
+// startInternal initializes the core GSDK functionality
+func (i *internalGsdk) startInternal() {
+	i.startInternalMutex.Lock()
+	defer i.startInternalMutex.Unlock()
+	if i.startInternalRan {
+		return
+	}
+
+	i.currentGameState = GameStateInitializing
+	var err error
+	if i.configuration == nil {
+		i.configuration, err = i.getConfiguration()
+		if err != nil {
+			log.Fatal(err)
+		}
+	}
+
+	if i.configMap == nil {
+		i.configMap = i.createConfigMap(i.configuration)
+	}
+
+	heartbeatEndpoint := i.configMap[HeartbeatEndpointKey]
+	serverId := i.configMap[ServerIdKey]
+
+	initializeLogger(i.configMap[LogFolderKey])
+
+	logInfo(fmt.Sprintf("VM Agent Endpoint %s", heartbeatEndpoint))
+	logInfo(fmt.Sprintf("Instance Id %s", serverId))
+
+	// run heartbeats on a different goroutine
+	go i.heartbeat(fmt.Sprintf("http://%s/v1/sessionHosts/%s/heartbeats", heartbeatEndpoint, serverId))
+	i.startInternalRan = true
+}
+
+func (i *internalGsdk) heartbeat(heartbeatEndpoint string) {
+	// if we get a signal for a state transition, we'll heartbeat
+	// else, we just heartbeat every second
+	for {
+		select {
+		case <-i.signalHeartbeatEvent:
+			if i.debugLogs {
+				logInfo("State transition signaled an early heartbeat")
+			}
+			i.sendHeartbeat(heartbeatEndpoint)
+		case <-time.After(time.Second * 1):
+			i.sendHeartbeat(heartbeatEndpoint)
+		}
+	}
+}
+
+func (i *internalGsdk) sendHeartbeat(heartbeatEndpoint string) {
+	gameHealth := false
+	if i.healthCallback != nil {
+		gameHealth = i.healthCallback()
+	}
+	currentGameHealth := "Healthy"
+	if !gameHealth {
+		currentGameHealth = "Unhealthy"
+	}
+	hrr := HeartbeatRequest{
+		CurrentGameState:  i.currentGameState,
+		CurrentGameHealth: currentGameHealth,
+		CurrentPlayers:    i.connectedPlayers,
+	}
+
+	postBody, _ := json.Marshal(hrr)
+	requestBody := bytes.NewBuffer(postBody)
+
+	resp, err := http.Post(heartbeatEndpoint, "application/json", requestBody)
+	if err != nil {
+		logWarn(fmt.Sprintf("Error sending heartbeat %s", err))
+		return
+	}
+
+	defer resp.Body.Close()
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		logWarn(fmt.Sprintf("Error reading heartbeat response %s", err))
+	}
+	hbr := HeartbeatResponse{}
+	if err := json.Unmarshal(body, &hbr); err != nil {
+		logWarn(fmt.Sprintf("Error parsing heartbeat response %s", err))
+	}
+
+	i.updateStateFromHeartbeat(&hbr)
```

Comment:
Looking at the Java GSDK here (https://github.com/PlayFab/gsdk/blob/34e0dae14c5f51fe28acaaf2ff60e2b1d53af07b/java/gameserverSDK/src/main/java/com/microsoft/azure/gaming/HeartbeatThread.java#L221) it looks like we check to see if we're in the `Terminating` state before we process the incoming heartbeat.  If we are in that state then we immediately send a heartbeat to the Agent saying that we're in the `Terminated` state and then halt further processing.  I haven't read this entire PR yet, but I don't see anything in this GSDK that sends the `Terminated` heartbeat to the Agent.

Reply:
hm, that's interesting - I don't think we send the `Terminated` ever on the C# GSDK. We'll need to check this and see what's going on